src/tests/Loader/CollectibleAssemblies/AssemblyOnDisk/Program.cs

```diff
@@ -0,0 +1,90 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.IO;
+using System.Runtime.CompilerServices;
+using System.Runtime.Loader;
+using Xunit;
+
+public class Program
+{
+    [Fact]
+    public static void EntryPoint()
+    {
+        string directoryPath = Path.Combine(AppContext.BaseDirectory, "ToDelete");
+        string originalAssemblyPath = typeof(Program).Assembly.Location;
+        string newAssemblyPath = Path.Combine(directoryPath, Path.GetFileName(originalAssemblyPath));
+
+        // If the directory already exists, delete it
+        if (Directory.Exists(directoryPath))
+        {
+            Console.WriteLine("Temp directory already exists, deleting...");
+            Directory.Delete(directoryPath, true);
+        }
+
+        // Create a directory to copy the assembly to
+        Directory.CreateDirectory(directoryPath);
+        try
+        {
+            File.Copy(originalAssemblyPath, newAssemblyPath);
+
+            UnloadableAssemblyContext assemblyContext = new();
+            assemblyContext.RunWithAssemblyLoadContext(
+                context =>
+                {
+                    context.LoadFromAssemblyPath(newAssemblyPath);
+                });
+
+            assemblyContext.Unload();
+        }
+        finally
+        {
+            Directory.Delete(directoryPath, recursive: true);
+        }
+
+        Assert.False(Directory.Exists(directoryPath));
+    }
+
+    class UnloadableAssemblyContext
+    {
+        private readonly WeakReference _weakAssemblyLoadContextReference;
+
+        private AssemblyLoadContext? _assemblyLoadContext;
+
+        public UnloadableAssemblyContext()
```

Comment:
This need to be marked as no-inlining too.

It is valid for JIT can inline this method and leave a refence to collectible assembly load context in a local that is kept alive till the end of `EntryPoint` method.

The usual pattern for these tests is to have a single NoInline method, do everything that touches the collectible  stuff inside the method (including creation of unloadable ALC), and check that everything was collected after that method returns.